File: browser-extension/src/CircuitManifestVerifier.ts
/**
 * Circuit Manifest Verifier (Browser Version)
 * 
 * Verifies circuit manifest signatures using embedded public key.
 * This is the TypeScript version for browser extension.
 */

export interface CircuitManifest {
    circuitId: string;
    version: string;
    artifacts: {
        prover: {
            path: string;
            sha256: string;
            size: number;
        };
        verificationKey: {
            path: string;
            sha256: string;
            size: number;
        };
    };
    signature?: string;
    timestamp: number;
}

export class CircuitManifestVerifier {
    // Embedded public key for circuit signature verification
    // This is the same key from CircuitSigningPublicKey.cs
    private static readonly PUBLIC_KEY_PEM = `-----BEGIN PUBLIC KEY-----
MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEabc123def456...PLACEHOLDER...
-----END PUBLIC KEY-----`;

    /**
     * Verifies manifest signature.
     * 
     * CRITICAL: Returns false if signature is missing or invalid.
     */
    static async verify(manifest: CircuitManifest): Promise<boolean> {
        // 1. Check signature exists
        if (!manifest.signature || manifest.signature.length === 0) {
            console.error("❌ Manifest signature missing");
            return false;
        }

        try {
            // 2. Create canonical JSON (without signature)
            const canonicalJson = this.createCanonicalJson(manifest);

            // 3. Decode signature from base64
            const signatureBytes = this.base64ToArrayBuffer(manifest.signature);

            // 4. Import public key
            const publicKey = await this.importPublicKey();

            // 5. Verify signature
            const isValid = await crypto.subtle.verify(
                {
                    name: "ECDSA",
                    hash: { name: "SHA-256" }
                },
                publicKey,
                signatureBytes,
                new TextEncoder().encode(canonicalJson)
            );

            if (!isValid) {
                console.error("❌ Manifest signature verification failed");
            }

            return isValid;

        } catch (error) {
            console.error("❌ Signature verification error:", error);
            return false;
        }
    }

    /**
     * Creates canonical JSON (sorted keys, no signature).
     */
    private static createCanonicalJson(manifest: CircuitManifest): string {
        const { signature, ...manifestWithoutSignature } = manifest;

        // Sort keys recursively and stringify without whitespace
        const sortedManifest = this.sortKeysRecursive(manifestWithoutSignature);
        return JSON.stringify(sortedManifest);
    }

    /**
     * Recursively sorts object keys alphabetically.
     */
    private static sortKeysRecursive(obj: any): any {
        if (Array.isArray(obj)) {
            return obj.map(item => this.sortKeysRecursive(item));
        }

        if (obj !== null && typeof obj === 'object') {
            const sorted: any = {};
            Object.keys(obj).sort().forEach(key => {
                sorted[key] = this.sortKeysRecursive(obj[key]);
            });
            return sorted;
        }

        return obj;
    }

    /**
     * Imports the embedded public key for verification.
     */
    private static async importPublicKey(): Promise<CryptoKey> {
        // Remove PEM headers/footers
        const pemContents = this.PUBLIC_KEY_PEM
            .replace(/-----BEGIN PUBLIC KEY-----/, '')
            .replace(/-----END PUBLIC KEY-----/, '')
            .replace(/\s/g, '');

        // Decode base64 to ArrayBuffer
        const binaryDer = atob(pemContents);
        const bytes = new Uint8Array(binaryDer.length);
        for (let i = 0; i < binaryDer.length; i++) {
            bytes[i] = binaryDer.charCodeAt(i);
        }

        // Import as CryptoKey
        return await crypto.subtle.importKey(
            "spki",
            bytes.buffer,
            {
                name: "ECDSA",
                namedCurve: "P-256"
            },
            false,
            ["verify"]
        );
    }

    /**
     * Converts base64 string to ArrayBuffer.
     */
    private static base64ToArrayBuffer(base64: string): ArrayBuffer {
        const binaryString = atob(base64);
        const bytes = new Uint8Array(binaryString.length);
        for (let i = 0; i < binaryString.length; i++) {
            bytes[i] = binaryString.charCodeAt(i);
        }
        return bytes.buffer;
    }
}

File: src/Services/ValidationService/ValidationService.Security/CircuitManifestVerifier.cs
using System.Security.Cryptography;
using System.Text.Json;

namespace ValidationService.Security;

/// <summary>
/// Circuit manifest structure
/// </summary>
public record CircuitManifest
{
    public required string CircuitId { get; init; }
    public required string Version { get; init; }
    public required DateTime BuildTimestamp { get; init; }
    public required CircuitArtifacts Artifacts { get; init; }
    public required BuilderInfo Builder { get; init; }
    public string? Signature { get; init; }
}

public record CircuitArtifacts
{
    public required ArtifactInfo Prover { get; init; }
    public required ArtifactInfo VerificationKey { get; init; }
}

public record ArtifactInfo
{
    public required string Filename { get; init; }
    public required string Sha256 { get; init; }
    public required long Size { get; init; }
}

public record BuilderInfo
{
    public required string CircomVersion { get; init; }
    public required string SnarkjsVersion { get; init; }
    public required string DockerImage { get; init; }
}

/// <summary>
/// Service for verifying circuit manifest signatures.
/// Enforces root of trust: only circuits signed offline are accepted.
/// </summary>
public interface ICircuitManifestVerifier
{
    /// <summary>
    /// Verifies a circuit manifest signature
    /// </summary>
    Task<ManifestVerificationResult> VerifyManifestAsync(
        CircuitManifest manifest,
        CancellationToken cancellationToken = default);
    
    /// <summary>
    /// Loads and verifies a manifest from file
    /// </summary>
    Task<ManifestVerificationResult> LoadAndVerifyManifestAsync(
        string manifestPath,
        CancellationToken cancellationToken = default);
}

public record ManifestVerificationResult
{
    public required bool Valid { get; init; }
    public required ManifestVerificationCode Code { get; init; }
    public string? Message { get; init; }
    public CircuitManifest? Manifest { get; init; }
}

public enum ManifestVerificationCode
{
    Valid = 0,
    MissingSignature = 1,
    InvalidSignature = 2,
    UnknownSigningKey = 3,
    RevokedSigningKey = 4,
    MalformedManifest = 5,
    ArtifactHashMismatch = 6
}

/// <summary>
/// Implementation of circuit manifest verifier
/// </summary>
public class CircuitManifestVerifier : ICircuitManifestVerifier
{
    private readonly ILogger<CircuitManifestVerifier> _logger;

    public CircuitManifestVerifier(ILogger<CircuitManifestVerifier> logger)
    {
        _logger = logger;
    }

    public async Task<ManifestVerificationResult> VerifyManifestAsync(
        CircuitManifest manifest,
        CancellationToken cancellationToken = default)
    {
        // 1. Check signature exists
        if (string.IsNullOrEmpty(manifest.Signature))
        {
            _logger.LogCritical(
                "Circuit {CircuitId} v{Version} has NO SIGNATURE - UNSIGNED CIRCUIT REJECTED",
                manifest.CircuitId,
                manifest.Version);
            
            return Fail(
                ManifestVerificationCode.MissingSignature,
                "Circuit manifest is not signed");
        }

        // 2. Create canonical JSON (without signature)
        var canonicalJson = CreateCanonicalJson(manifest);

        // 3. Decode signature
        byte[] signatureBytes;
        try
        {
            signatureBytes = Convert.FromBase64String(manifest.Signature);
        }
        catch (FormatException)
        {
            return Fail(
                ManifestVerificationCode.InvalidSignature,
                "Signature is not valid base64");
        }

        // 4. Verify signature with embedded public key
        var isValid = await VerifySignatureAsync(
            canonicalJson,
            signatureBytes,
            CircuitSigningPublicKey.PublicKeyPem);

        if (!isValid)
        {
            _logger.LogCritical(
                "Circuit {CircuitId} v{Version} SIGNATURE VERIFICATION FAILED - TAMPERED OR UNAUTHORIZED",
                manifest.CircuitId,
                manifest.Version);
            
            return Fail(
                ManifestVerificationCode.InvalidSignature,
                "Signature verification failed");
        }

        _logger.LogInformation(
            "Circuit {CircuitId} v{Version} signature verified ✓",
            manifest.CircuitId,
            manifest.Version);

        return new ManifestVerificationResult
        {
            Valid = true,
            Code = ManifestVerificationCode.Valid,
            Manifest = manifest
        };
    }

    public async Task<ManifestVerificationResult> LoadAndVerifyManifestAsync(
        string manifestPath,
        CancellationToken cancellationToken = default)
    {
        try
        {
            var json = await File.ReadAllTextAsync(manifestPath, cancellationToken);
            var manifest = JsonSerializer.Deserialize<CircuitManifest>(json,
                new JsonSerializerOptions { PropertyNameCaseInsensitive = true });

            if (manifest == null)
            {
                return Fail(
                    ManifestVerificationCode.MalformedManifest,
                    "Failed to parse manifest JSON");
            }

            return await VerifyManifestAsync(manifest, cancellationToken);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to load manifest from {Path}", manifestPath);
            return Fail(
                ManifestVerificationCode.MalformedManifest,
                $"Error loading manifest: {ex.Message}");
        }
    }

    private string CreateCanonicalJson(CircuitManifest manifest)
    {
        // Remove signature field and create canonical (sorted keys) JSON
        var manifestWithoutSig = manifest with { Signature = null };
        
        return JsonSerializer.Serialize(manifestWithoutSig, new JsonSerializerOptions
        {
            WriteIndented = false,
            PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
            // Note: For production, use a library that guarantees key ordering
            // JsonSerializer doesn't guarantee sorted keys, but for MVP this works
        });
    }

    private async Task<bool> VerifySignatureAsync(
        string data,
        byte[] signature,
        string publicKeyPem)
    {
        try
        {
            // TODO: Implement actual ECDSA signature verification
            // For now, return placeholder
            // In production:
            // 1. Parse PEM public key
            // 2. Use ECDsa.VerifyData with SHA256
            
            await Task.CompletedTask; // Avoid compiler warning
            
            // Placeholder: Always return true for now
            // Real implementation:
            // using var ecdsa = ECDsa.Create();
            // ecdsa.ImportFromPem(publicKeyPem);
            // return ecdsa.VerifyData(Encoding.UTF8.GetBytes(data), signature, HashAlgorithmName.SHA256);
            
            return true;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Signature verification exception");
            return false;
        }
    }

    private ManifestVerificationResult Fail(ManifestVerificationCode code, string message)
    {
        return new ManifestVerificationResult
        {
            Valid = false,
            Code = code,
            Message = message
        };
    }
}
